tensorflow_addons/optimizers/discriminative_layer_training_test.py

```diff
@@ -0,0 +1,399 @@
+
+        self._assert_training_losses_are_close(model, model_lr)
+
+    @test_utils.run_distributed(2)
+    def _test_sub_layers_keep_lr_mult(self, model_fn, loss, opt):
+        """confirm that model trains with lower lr on specific layer
+        while a different lr_mult is applied everywhere else
+        also confirms that sub layers with an lr mult do not get overridden
+        """
+
+        learning_rate = 0.01
+        model_lr = model_fn()
+
+        # we set model to lrmult 0 and layer one to lrmult 5
+        # if layer one is trainable, then the loss should decrease
+        model_lr.lr_mult = 0.00
+        model_lr.layers[-1].lr_mult = 3
+
+        d_opt = DiscriminativeLayerOptimizer(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        loss_values = get_losses(_get_train_results(model_lr, epochs=5))
+        self.assertLess(loss_values[-1], loss_values[0])
+
+    @test_utils.run_distributed(2)
+    def _test_variables_get_assigned(self, model_fn, loss, opt):
+        """confirm that variables do get an lr_mult attribute and that they get the correct one
+        """
+        learning_rate = 0.01
+        model_lr = model_fn()
+
+        # set lr mults
+        model_lr.layers[0].lr_mult = 0.3
+        model_lr.layers[0].layers[-1].lr_mult = 0.1
+        model_lr.layers[-1].lr_mult = 0.5
+
+        d_opt = DiscriminativeLayerOptimizer(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        # we expect trainable vars at 0.3 to be reduced by the amount at 0.1
+        # this tests that the 0.3 lr mult does not override the 0.1 lr mult
+        self.assertEqual(
+            len(model_lr.layers[0].trainable_variables)
+            - len(model_lr.layers[0].layers[-1].trainable_variables),
+            len([var for var in model_lr.trainable_variables if var.lr_mult == 0.3]),
+        )
+
+        # we expect trainable vars of model with lr_mult 0.1 to equal trainable vars of that layer
+        self.assertEqual(
+            len(model_lr.layers[0].layers[-1].trainable_variables),
+            len([var for var in model_lr.trainable_variables if var.lr_mult == 0.1]),
+        )
+
+        # same logic as above
+        self.assertEqual(
+            len(model_lr.layers[-1].trainable_variables),
+            len([var for var in model_lr.trainable_variables if var.lr_mult == 0.5]),
+        )
+
+    def _run_tests_in_notebook(self):
+        for name, method in DiscriminativeLearningTest.__dict__.items():
+            if callable(method) and name[:4] == "test":
+                print("running test %s" % name)
+                method(self)
+
+
+def test_wrap(method, **kwargs):
+    # wrap every test to run in graph and eager
+
+    @test_utils.run_in_graph_and_eager_modes
+    def test(self):
+        return method(self, **kwargs)
+
+    return test
```

Comment:
Not needed. See `test_utils.run_all_in_graph_and_eager_modes`. It's a class wrapper, use it on `DiscriminativeLearningTest`.

Reply:
I believe the class wrapper checks for the the test prefix so I wasn't sure if this would interfere with generate_tests(). Because the base tests are prefixed with _test_dotest and then generated using the generator as test_dotest.

Reply:
I guess my question here is whether or not test_utils.run_all_in_graph_and_eager_modes modifies the attributes when the class is defined or when the class is instantiated?